src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -0,0 +1,317 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.SQLite.v2.Interop;
+using Microsoft.CodeAnalysis.Storage;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SQLite.v2
+{
+    internal partial class SQLitePersistentStorage
+    {
+        /// <summary>
+        /// Abstracts out access to specific tables in the DB.  This allows us to share overall
+        /// logic around cancellation/pooling/error-handling/etc, while still hitting different
+        /// db tables.
+        /// </summary>
+        private abstract class Accessor<TKey, TWriteQueueKey, TDatabaseId>
+        {
+            protected readonly SQLitePersistentStorage Storage;
+
+            // Cache the statement strings we want to execute per accessor.  This way we avoid
+            // allocating these strings each time we execute a command.  We also cache the prepared
+            // statements (at the connection level) we make for each of these strings.  That way we
+            // only incur the parsing cost once. After that, we can use the same prepared statements
+            // and just bind the appropriate values it needs into it.
+            //
+            // values like 0, 1, 2 in the name are the `?`s in the sql string that will need to be
+            // bound to runtime values appropriately when executed.
+
+            private readonly string _select_rowid_from_main_table_where_0;
+            private readonly string _select_rowid_from_writecache_table_where_0;
+            private readonly string _insert_or_replace_into_writecache_table_values_0_1_2;
+            private readonly string _delete_from_writecache_table;
+            private readonly string _insert_or_replace_into_main_table_select_star_from_writecache_table;
+
+            public Accessor(SQLitePersistentStorage storage)
+            {
+                var main = Database.Main.GetName();
+                var writeCache = Database.WriteCache.GetName();
+
+                Storage = storage;
+                _select_rowid_from_main_table_where_0 = $@"select rowid from {main}.{DataTableName} where ""{DataIdColumnName}"" = ?";
+                _select_rowid_from_writecache_table_where_0 = $@"select rowid from {writeCache}.{DataTableName} where ""{DataIdColumnName}"" = ?";
+                _insert_or_replace_into_writecache_table_values_0_1_2 = $@"insert or replace into {writeCache}.{DataTableName}(""{DataIdColumnName}"",""{ChecksumColumnName}"",""{DataColumnName}"") values (?,?,?)";
+                _delete_from_writecache_table = $"delete from {writeCache}.{DataTableName};";
+                _insert_or_replace_into_main_table_select_star_from_writecache_table = $"insert or replace into {main}.{DataTableName} select * from {writeCache}.{DataTableName};";
```

Comment:
SQLite doesn't let us batch both as a single command?

Reply:
90% certain the answer is know.  Tagging @ericsink To see if he has thoughts.  My understanding @ericsink is that i would create a transaction and just execute both of these statements.  from sqlite's perspective, that's "the right way" to do things here.